### PR TITLE
NIAD-383: inbound escaped characters

### PR DESCRIPTION
**Pull request by jaroslaw-wojciechowski**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java [hunk 1]
```diff
@@ -2,22 +2,22 @@
...
     }
 
-    public static String[] byNewLine(String input) {
-        return input.split("\\n");
+    /**
+     * Matches given delimiter NOT preceded by a odd occurrence of question mark
+     * Works for 1 & 3, Lookbehinds must be fixed-width.
+     */
+    private static String[] splitByDelimiter(String input, String delimiter) {
+        return input.split(String.format("(?<![^\\?]\\?|[^\\?]\\?{3})%s", delimiter));
```
Comment (jaroslaw-wojciechowski) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: regex 'Lookbehinds' has some limitations. I haven't found anything better.
SplitTest.java shows which cases are covered.

File: src/intTest/resources/inbound_uat_data/rejection/full-character-set.fhir.json [hunk 2]
```diff
@@ -1,6 +1,10 @@
 {
   "resourceType": "Parameters",
   "parameter": [
+    {
```
Comment (bartek-sarul) [inline src/intTest/resources/inbound_uat_data/rejection/full-character-set.fhir.json:4]: Why the order has changed?

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/SplitTest.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static uk.nhs.digital.nhsconnect.nhais.model.edifact.message.Split.byColon;
+import static uk.nhs.digital.nhsconnect.nhais.model.edifact.message.Split.byPlus;
+import static uk.nhs.digital.nhsconnect.nhais.model.edifact.message.Split.bySegmentTerminator;
+
+class SplitTest {
+    @Test
+    public void When_SplittingBySegmentTerminator_Expect_CorrectResult() {
+        assertEquals("asdf", bySegmentTerminator("asdf'test-string")[0]);
```
Comment (bartek-sarul) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/SplitTest.java]: Use AssertJ

**Review by bartek-sarul — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java (hunk 1 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: Why did the much simpler `"((?<!\\?)')"` work?

Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: It probably doesn't work for multiple ??? before the delimiter

Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: What if... we took input and replaced all `??` by for example `@@` and then applied regex for single `?` and then changed `@@` back to `??` ?
Or just parse those escape characters and then apply regex?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java [hunk 4]
```diff
@@ -1,23 +1,50 @@
...
+                adjacentEscCharsCount = 0;
+            }
 
+            sb.append(c);
+        }
+        parts.add(sb.toString());
+
+        String[] array = new String[parts.size()];
+        parts.toArray(array);
+        return array;
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: (Optional) or just `return parts.toArray(String[]::new)`

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java [hunk 5]
```diff
@@ -1,23 +1,50 @@
...
-    public static String[] bySegmentTerminator(String input) {
-        return input.split("((?<!\\?)')");
+    public static String[] byColon(String input) {
+        return splitString(input, SUB_FIELD_TERMINATOR);
     }
 
-    public static String[] byNewLine(String input) {
-        return input.split("\\n");
-    }
+    static String[] splitString(String input, char separator) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/message/Split.java]: private?

**Review by grzegorz-robak — DISMISSED** [review]

File: src/intTest/resources/inbound_uat_data/rejection/full-character-set.fhir.json (hunk 2 above)
Comment (bartek-sarul) [reply src/intTest/resources/inbound_uat_data/rejection/full-character-set.fhir.json:4]: Got it. It was `ignore` before and once unlocked you had to adjust it

**Review by grzegorz-robak — APPROVED** [review]

**Pull request merged**